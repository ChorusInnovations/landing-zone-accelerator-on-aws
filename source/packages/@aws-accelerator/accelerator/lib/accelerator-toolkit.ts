/**
 *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as config from '@aws-accelerator/config';
import { AssumeProfilePlugin } from '@aws-cdk-extensions/cdk-plugin-assume-role';
import { SdkProvider } from 'aws-cdk/lib/api/aws-auth';
import { Bootstrapper, BootstrapSource } from 'aws-cdk/lib/api/bootstrap';
import { CloudFormationDeployments } from 'aws-cdk/lib/api/cloudformation-deployments';
import { StackSelector } from 'aws-cdk/lib/api/cxapp/cloud-assembly';
import { CloudExecutable } from 'aws-cdk/lib/api/cxapp/cloud-executable';
import { execProgram } from 'aws-cdk/lib/api/cxapp/exec';
import { ToolkitInfo } from 'aws-cdk/lib/api/toolkit-info';
import { CdkToolkit } from 'aws-cdk/lib/cdk-toolkit';
import { PluginHost } from 'aws-cdk/lib/plugin';
import { Command, Configuration } from 'aws-cdk/lib/settings';
import console from 'console';
import * as fs from 'fs';
import mri from 'mri';
import path from 'path';
import process from 'process';
import { Stage } from './stages';

/**
 * Wrapper around the CdkToolkit. The Accelerator defines this wrapper to add
 * the following functionality:
 *
 * - x
 * - y
 * - z
 */
export class AcceleratorToolkit {
  /**
   *
   * @returns
   */
  static async cli(): Promise<void> {
    const usage = `Usage: cdk.ts <command> --stage STAGE --config-dir CONFIG_DIRECTORY [--account ACCOUNT] [--region REGION] [--parallel]`;
    const args = mri(process.argv.slice(2), {
      boolean: ['parallel'],
      string: ['account'],
      alias: {
        c: 'config-dir',
        s: 'stage',
        a: 'account',
        r: 'region',
      },
      default: {
        parallel: false,
      },
    });

    const commands = args['_'];
    // const parallel = args['parallel'];
    const configDir = args['config-dir'];
    const stage = args['stage'];
    const account = args['account'];
    const region = args['region'];

    //
    // Validate args: must specify a command
    //
    if (commands.length === 0) {
      console.log('<command> not set');
      throw new Error(usage);
    }

    //
    // Validate args: verify command against our sub-list
    //
    const supportedCommands: string[] = [
      Command.BOOTSTRAP,
      Command.DEPLOY,
      Command.DESTROY,
      Command.LIST,
      Command.LS,
      Command.SYNTH,
      Command.SYNTHESIZE,
    ];
    if (!supportedCommands.includes(commands[0])) {
      throw new Error(`Invalid command: ${commands[0]}`);
    }

    //
    // Validate args: verify stage if not bootstrap or list
    //
    if (!Object.values(Stage).includes(stage) && commands[0] !== 'bootstrap') {
      throw new Error(`Invalid stage: ${stage}`);
    }

    //
    // Validate args: verify config directory
    //
    if (config === undefined || !fs.existsSync(configDir)) {
      console.log(`Invalid --config ${configDir}`);
      throw new Error(usage);
    }

    //
    // Load Plugins
    //
    const assumeRolePlugin = new AssumeProfilePlugin({
      // TODO: Read this from arg
      assumeRoleName: 'AWSControlTowerExecution',
      assumeRoleDuration: 3600,
    });
    assumeRolePlugin.init(PluginHost.instance);

    //
    // When an account and region is specified, execute as single stack
    //
    if (account || region) {
      if (account && region === undefined) {
        throw new Error(`Account set to ${account}, but region is undefined`);
      }
      if (region && account === undefined) {
        throw new Error(`Region set to ${region}, but region is undefined`);
      }
      return await AcceleratorToolkit.execute(commands[0], account, region, stage);
    }

    //
    // Read in all Accelerator Configuration files here, then pass the objects
    // to the stacks that need them
    //
    const organizationConfig = await config.loadOrganizationConfig(configDir);
    console.log(organizationConfig);

    // TODO: And Environment variables to enable debug logs

    //
    // Loop through all accounts and regions and execute commands
    //
    // TODO: Add parallel support
    // TODO: Change config to not include account numbers, need to pull from orgs
    for (const account in organizationConfig['accounts']) {
      for (const region of organizationConfig['enabled-regions']) {
        await AcceleratorToolkit.execute(commands[0], account, region, stage);
      }
    }
  }

  /**
   * Accelerator customized execution of the CDKToolkit based on
   * aws-cdk/packages/aws-cdk/bin/cdk.ts
   *
   *
   * @param command
   * @param account
   * @param region
   * @param stage
   *
   * @return Promise<void>
   *
   * @see aws-cdk/packages/aws-cdk/bin/cdk.ts
   */
  static async execute(command: string, account: string, region: string, stage: string): Promise<void> {
    console.log(`Executing cdk ${command} for aws://${account}/${region}`);

    const configuration = new Configuration({
      commandLineArguments: {
        _: [Command.BOOTSTRAP, ...[]],
        pathMetadata: false,
        assetMetadata: false,
        versionReporting: false,
        output: path.join('cdk.out', account, region),
        context: [`account=${account}`, `region=${region}`, `stage=${stage}`],
      },
    });
    await configuration.load();

    const sdkProvider = await SdkProvider.withAwsCliCompatibleDefaults({
      profile: configuration.settings.get(['profile']),
    });

    const cloudFormation = new CloudFormationDeployments({ sdkProvider });

    const cloudExecutable = new CloudExecutable({
      configuration,
      sdkProvider,
      synthesizer: execProgram,
    });

    const toolkitStackName: string = ToolkitInfo.determineName(configuration.settings.get(['toolkitStackName']));

    const cli = new CdkToolkit({
      cloudExecutable,
      cloudFormation,
      configuration,
      sdkProvider,
    });

    const selector: StackSelector = {
      patterns: [],
    };

    switch (command) {
      case Command.BOOTSTRAP:
        const source: BootstrapSource = { source: 'default' };
        const bootstrapper = new Bootstrapper(source);
        const environments = [`aws://${account}/${region}`];
        await cli.bootstrap(environments, bootstrapper, {
          toolkitStackName,
          parameters: {
            bucketName: configuration.settings.get(['toolkitBucket', 'bucketName']),
            kmsKeyId: configuration.settings.get(['toolkitBucket', 'kmsKeyId']),
          },
        });
        break;
      case Command.DEPLOY:
        await cli.deploy({
          selector,
          toolkitStackName,
        });
        break;
      case Command.DESTROY:
        await cli.destroy({
          selector,
          exclusively: false,
          force: true,
        });
        break;
      case Command.LIST:
      case Command.LS:
        await cli.list([], { long: false });
        break;
      case Command.SYNTHESIZE:
      case Command.SYNTH:
        await cli.synth([], false, true);
        break;
      default:
        throw new Error(`Unsupported command: ${command}`);
    }
  }
}
