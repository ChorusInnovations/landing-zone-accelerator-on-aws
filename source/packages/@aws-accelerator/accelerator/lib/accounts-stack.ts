/**
 *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

/**
 * @module
 * Module comment
 */

import * as cdk from '@aws-cdk/core';
import * as ssm from '@aws-cdk/aws-ssm';

export interface AccountsStackProps extends cdk.StackProps {
  stage: string;
}

export class AccountsStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props: AccountsStackProps) {
    super(scope, id, props);

    new ssm.StringParameter(this, 'Parameter', {
      parameterName: `/accelerator/accounts-stack/${props.stage}`,
      stringValue: 'value',
    });
  }
}
