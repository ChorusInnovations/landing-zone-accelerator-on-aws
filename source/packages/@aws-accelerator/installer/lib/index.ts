export * from './pipeline';
export * from './installer-stack';
export * from './config-repository';
