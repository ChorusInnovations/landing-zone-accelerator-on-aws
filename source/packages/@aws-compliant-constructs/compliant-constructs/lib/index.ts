export * from './secure-s3-bucket';
