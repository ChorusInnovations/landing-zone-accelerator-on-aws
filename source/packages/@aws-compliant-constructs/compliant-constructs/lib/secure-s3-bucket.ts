import * as iam from '@aws-cdk/aws-iam';
import * as kms from '@aws-cdk/aws-kms';
import * as s3 from '@aws-cdk/aws-s3';
import * as cdk from '@aws-cdk/core';

export enum BucketAccessType {
  READONLY = 'readonly',
  WRITEONLY = 'writeonly',
  READWRITE = 'readwrite',
}

/**
 * Construction properties for a Secure S3 Bucket object.
 */
export interface SecureS3BucketProps {
  /**
   * Physical name of this bucket.
   *
   * @default - Assigned by CloudFormation (recommended).
   */
  readonly s3BucketName?: string;
  /**
   * Policy to apply when the bucket is removed from this stack.
   *
   * @default - The bucket will be orphaned.
   */
  readonly s3RemovalPolicy?: cdk.RemovalPolicy;
  /**
   * The name of the alias.
   */
  readonly kmsAliasName: string;
  /**
   * A description of the key.
   *
   * Use a description that helps your users decide
   * whether the key is appropriate for a particular task.
   *
   */
  readonly kmsDescription: string;

  /**
   * @optional
   * A list of AWS principals and access type the bucket to grant
   * principal should be a valid AWS resource principal like for AWS Macie it should be macie.amazonaws.com
   * accessType should be any of these possible values BucketAccessType.READONLY, BucketAccessType.WRITEONLY, & and BucketAccessType.READWRITE
   */
  readonly awsPrincipalAccesses?: { principalAccesses: [{ principal: string; accessType: string }] };
}

/**
 * Defines a Secure S3 Bucket object. By default a KMS CMK is generated and
 * associated to the bucket.
 */
export class SecureS3Bucket extends cdk.Construct {
  private readonly bucket: s3.Bucket;
  private readonly cmk: kms.Key;

  constructor(scope: cdk.Construct, id: string, props: SecureS3BucketProps) {
    super(scope, id);

    this.cmk = new kms.Key(this, 'Cmk', {
      enableKeyRotation: true,
      description: props.kmsDescription,
    });
    this.cmk.addAlias(props.kmsAliasName);

    this.bucket = new s3.Bucket(this, 'Resource', {
      encryption: s3.BucketEncryption.KMS,
      encryptionKey: this.cmk,
      removalPolicy: cdk.RemovalPolicy.RETAIN,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      bucketName: props.s3BucketName,
      versioned: true,
      objectOwnership: s3.ObjectOwnership.BUCKET_OWNER_PREFERRED,
    });
    this.bucket.addToResourcePolicy(
      new iam.PolicyStatement({
        sid: 'deny-non-encrypted-object-uploads',
        effect: iam.Effect.DENY,
        actions: ['s3:PutObject'],
        resources: [this.bucket.arnForObjects('*')],
        principals: [new iam.AnyPrincipal()],
        conditions: {
          StringNotEquals: {
            's3:x-amz-server-side-encryption': 'aws:kms',
          },
        },
      }),
    );
    this.bucket.addToResourcePolicy(
      new iam.PolicyStatement({
        sid: 'deny-insecure-connections',
        effect: iam.Effect.DENY,
        actions: ['s3:*'],
        resources: [this.bucket.arnForObjects('*')],
        principals: [new iam.AnyPrincipal()],
        conditions: {
          Bool: {
            'aws:SecureTransport': 'false',
          },
        },
      }),
    );

    // Add access policy for inputed AWS principal to the bucket
    // (add a `if` to make TS happy)
    if (props.awsPrincipalAccesses) {
      props.awsPrincipalAccesses.principalAccesses.forEach(input => {
        switch (input.accessType) {
          case BucketAccessType.READONLY:
            this.bucket.grantRead(new iam.ServicePrincipal(input.principal));
            break;
          case BucketAccessType.WRITEONLY:
            this.bucket.grantWrite(new iam.ServicePrincipal(input.principal));
            break;
          case BucketAccessType.READWRITE:
            this.bucket.grantReadWrite(new iam.ServicePrincipal(input.principal));
            break;
          default:
            throw new Error(`Invalid Access Type ${input.accessType} for ${input.principal} principal.`);
        }
      });
    }
  }

  public getS3Bucket(): s3.IBucket {
    return this.bucket;
  }

  protected override validate(): string[] {
    const errors: string[] = [];

    // errors.push('some error');

    return errors;
  }
}
