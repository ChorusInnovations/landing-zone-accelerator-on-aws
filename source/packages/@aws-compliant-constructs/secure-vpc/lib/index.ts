import * as cdk from '@aws-cdk/core';
import * as ssm from '@aws-cdk/aws-ssm';

export interface SecureVpcProps {
  // Define construct properties here
}

export class SecureVpc extends cdk.Construct {
  constructor(scope: cdk.Construct, id: string, props: SecureVpcProps = {}) {
    super(scope, id);

    // Define construct contents here
    new ssm.StringParameter(this, 'SecureVpcSsmTest', {
      allowedPattern: '.*',
      description: 'The value Foo',
      parameterName: 'FooParameter',
      stringValue: 'Foo',
    });
  }
}
