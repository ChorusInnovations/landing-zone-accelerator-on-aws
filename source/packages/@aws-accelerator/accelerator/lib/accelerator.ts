/**
 *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
import { AccountsConfig, GlobalConfig, OrganizationConfig, SecurityConfig } from '@aws-accelerator/config';
import { throttlingBackOff } from '@aws-accelerator/utils';
import { AssumeProfilePlugin } from '@aws-cdk-extensions/cdk-plugin-assume-role';
import { DescribeOrganizationCommand, OrganizationsClient, paginateListAccounts } from '@aws-sdk/client-organizations';
import { RequireApproval } from 'aws-cdk/lib/diff';
import { PluginHost } from 'aws-cdk/lib/plugin';
import { AcceleratorToolkit } from './toolkit';

export enum AcceleratorStage {
  PIPELINE = 'pipeline',
  /**
   * Validate Stage - Verify the configuration files and environment
   */
  VALIDATE = 'validate',
  ORGANIZATIONS = 'organizations',
  LOGGING = 'logging',
  /**
   * Accounts Stage - Handle all Organization and Accounts actions
   */
  ACCOUNTS = 'accounts',
  DEPENDENCIES = 'dependencies',
  SECURITY = 'security',
  OPERATIONS = 'operations',
  NETWORKING = 'networking',
  'SECURITY-AUDIT' = 'security-audit',
}

/**
 *
 */
export interface AcceleratorProps {
  command: string;
  configDirPath: string;
  parallel: boolean;
  stage: string;
  account: string;
  region: string;
  partition: string;
  requireApproval: RequireApproval;
}

/**
 * Wrapper around the CdkToolkit. The Accelerator defines this wrapper to add
 * the following functionality:
 *
 * - x
 * - y
 * - z
 */
export abstract class Accelerator {
  static isSupportedStage(stage: AcceleratorStage): boolean {
    return Object.values(AcceleratorStage).includes(stage);
  }

  /**
   *
   * @returns
   */
  static async run(props: AcceleratorProps): Promise<void> {
    //
    // When an account and region is specified, execute as single stack
    //
    if (props.account || props.region) {
      if (props.account && props.region === undefined) {
        throw new Error(`Account set to ${props.account}, but region is undefined`);
      }
      if (props.region && props.account === undefined) {
        throw new Error(`Region set to ${props.region}, but region is undefined`);
      }
      return await AcceleratorToolkit.execute({
        command: props.command,
        accountId: props.account,
        region: props.region,
        stage: props.stage,
        configDirPath: props.configDirPath,
        requireApproval: props.requireApproval,
      });
    }

    //
    // Read in all Accelerator Configuration files here, then pass the objects
    // to the stacks that need them. Exceptions are thrown if any of the
    // configuration files are malformed.
    //
    const globalConfig = GlobalConfig.load(props.configDirPath);
    const organizationsConfig = OrganizationConfig.load(props.configDirPath);
    const accountsConfig = AccountsConfig.load(props.configDirPath);
    const securityConfig = SecurityConfig.load(props.configDirPath);

    //
    // Load Plugins
    //
    const assumeRolePlugin = new AssumeProfilePlugin({
      // TODO: Read this from arg
      assumeRoleName: organizationsConfig['organizations-access-role'],
      assumeRoleDuration: 3600,
    });
    assumeRolePlugin.init(PluginHost.instance);

    //
    // NOTE: We do some early environment validation here before we kick off the
    //       CodePipeline that has a built in validation-stack
    //

    //
    // Verify AWS Organizations has been enabled
    //
    const organizationsClient = new OrganizationsClient({});
    await throttlingBackOff(() =>
      organizationsClient.send(new DescribeOrganizationCommand({})).catch(error => {
        if (error.name === 'AWSOrganizationsNotInUseException') {
          throw new Error(error.message);
        }
        throw new Error(error);
      }),
    );

    //
    // Create a dictionary of all AWS Account IDs.
    //
    // TODO: Add functionality to read in a map from file for disconnected regions
    //
    const accountIds: { [name: string]: string } = {};
    for await (const page of paginateListAccounts({ client: organizationsClient }, {})) {
      for (const account of page.Accounts ?? []) {
        if (account.Email && account.Id) {
          accountIds[account.Email] = account.Id;
        }
      }
    }

    //
    // Execute Bootstrap stacks for all identified accounts
    //
    if (props.command == 'bootstrap') {
      const trustedAccountId = accountIds[accountsConfig['mandatory-accounts'].management.email];
      for (const region of globalConfig['enabled-regions']) {
        for (const account of Object.values(accountsConfig['mandatory-accounts'])) {
          const accountId = accountIds[account.email];
          await AcceleratorToolkit.execute({
            command: props.command,
            accountId,
            region,
            partition: props.partition,
            trustedAccountId,
            requireApproval: props.requireApproval,
          });
        }
        for (const account of Object.values(accountsConfig['workload-accounts'])) {
          const accountId = accountIds[account.email];
          await AcceleratorToolkit.execute({
            command: props.command,
            accountId,
            region,
            partition: props.partition,
            trustedAccountId,
            requireApproval: props.requireApproval,
          });
        }
      }
      return;
    }

    // TODO: Need to decide the mandatory accounts for an accelerator --
    // Control Tower: To start a well-planned OU structure in your landing zone, AWS Control Tower
    // sets up a Security OU for you. This OU contains three shared accounts: the management
    // (primary) account, the log archive account, and the security audit account (also referred to
    // as the audit account).

    switch (props.stage) {
      case AcceleratorStage.VALIDATE:
      case AcceleratorStage.ACCOUNTS:
        break;
      case AcceleratorStage.ORGANIZATIONS:
        const managementAccount = accountsConfig['mandatory-accounts'].management;
        for (const region of globalConfig['enabled-regions']) {
          await AcceleratorToolkit.execute({
            command: props.command,
            accountId: accountIds[managementAccount.email],
            region: region,
            stage: props.stage,
            configDirPath: props.configDirPath,
            requireApproval: props.requireApproval,
          });
        }
        break;

      //
      // Apply these stacks to all account / regions. The contents of these stacks are dynamically
      // built from the inputted configuration files during stack construction
      //
      case AcceleratorStage.LOGGING:
      case AcceleratorStage.DEPENDENCIES:
      case AcceleratorStage.SECURITY:
        for (const region of globalConfig['enabled-regions'].filter(
          item => securityConfig.getExcludeRegions()?.indexOf(item) === -1,
        )) {
          for (const account of Object.values(accountsConfig['mandatory-accounts'])) {
            console.log(`${region} to be implemented on for ${account.email} account`);
            const accountId = accountIds[account.email];
            await AcceleratorToolkit.execute({
              command: props.command,
              accountId,
              region,
              stage: props.stage,
              configDirPath: props.configDirPath,
              requireApproval: props.requireApproval,
            });
          }
          for (const account of Object.values(accountsConfig['workload-accounts'])) {
            const accountId = accountIds[account.email];
            await AcceleratorToolkit.execute({
              command: props.command,
              accountId,
              region,
              stage: props.stage,
              configDirPath: props.configDirPath,
              requireApproval: props.requireApproval,
            });
          }
        }
        break;
      case AcceleratorStage.OPERATIONS:
      case AcceleratorStage.NETWORKING:
        for (const region of globalConfig['enabled-regions']) {
          for (const account of Object.values(accountsConfig['mandatory-accounts'])) {
            await AcceleratorToolkit.execute({
              command: props.command,
              accountId: accountIds[account.email],
              region,
              stage: props.stage,
              configDirPath: props.configDirPath,
              requireApproval: props.requireApproval,
            });
          }
          for (const account of Object.values(accountsConfig['workload-accounts'])) {
            await AcceleratorToolkit.execute({
              command: props.command,
              accountId: accountIds[account.email],
              region,
              stage: props.stage,
              configDirPath: props.configDirPath,
              requireApproval: props.requireApproval,
            });
          }
        }
        break;
      case AcceleratorStage['SECURITY-AUDIT']:
        const auditAccountName = securityConfig.getDelegatedAccountName();
        if (accountsConfig.accountExists(auditAccountName)) {
          for (const region of globalConfig['enabled-regions'].filter(
            item => securityConfig.getExcludeRegions()?.indexOf(item) === -1,
          )) {
            const auditAccountEmail = accountsConfig.getEmail(auditAccountName);
            const accountId = accountIds[auditAccountEmail];
            await AcceleratorToolkit.execute({
              command: props.command,
              accountId,
              region,
              stage: props.stage,
              configDirPath: props.configDirPath,
              requireApproval: props.requireApproval,
            });
          }
        } else {
          throw new Error(`Security audit delegated account ${auditAccountName} not found.`);
        }
        break;
      default:
        throw new Error(`Unknown stage: ${props.stage}`);
    }
  }
}
