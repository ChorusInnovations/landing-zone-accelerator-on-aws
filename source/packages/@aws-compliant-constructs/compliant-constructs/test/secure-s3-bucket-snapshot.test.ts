import { SynthUtils } from '@aws-cdk/assert';
import { test, describe, expect } from '@jest/globals';
import * as compliant_constructs from '../lib/secure-s3-bucket';
import * as TestConfig from './test-config';

describe('Secure S3 Bucket', () => {
  /**
   * Snapshot Test - Compliant Constructor Secure S3 Bucket
   */
  test.skip('Snapshot Test', () => {
    new compliant_constructs.SecureS3Bucket(TestConfig.stack, 'SecureBucket', TestConfig.secureBucketProps);
    expect(SynthUtils.toCloudFormation(TestConfig.stack)).toMatchSnapshot();
  });
});
