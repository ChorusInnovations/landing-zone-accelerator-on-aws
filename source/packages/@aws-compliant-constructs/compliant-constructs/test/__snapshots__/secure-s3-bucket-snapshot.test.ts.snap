// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Secure S3 Bucket Snapshot Test 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters00f2465b209575c44977ca8ff1c81d86d002849272ede5a8a90ac87dfff1e494ArtifactHashC2E6E6C0": Object {
      "Description": "Artifact hash for asset \\"00f2465b209575c44977ca8ff1c81d86d002849272ede5a8a90ac87dfff1e494\\"",
      "Type": "String",
    },
    "AssetParameters00f2465b209575c44977ca8ff1c81d86d002849272ede5a8a90ac87dfff1e494S3Bucket33B61C3E": Object {
      "Description": "S3 bucket for asset \\"00f2465b209575c44977ca8ff1c81d86d002849272ede5a8a90ac87dfff1e494\\"",
      "Type": "String",
    },
    "AssetParameters00f2465b209575c44977ca8ff1c81d86d002849272ede5a8a90ac87dfff1e494S3VersionKeyDE2514FF": Object {
      "Description": "S3 key for asset version \\"00f2465b209575c44977ca8ff1c81d86d002849272ede5a8a90ac87dfff1e494\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F": Object {
      "DependsOn": Array [
        "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
      ],
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W58",
              "reason": "Lambda functions deployed through CDK have the required permission to write CloudWatch Logs. It uses custom policy instead of AWSLambdaBasicExecutionRole with more tighter permissions.",
            },
            Object {
              "id": "W89",
              "reason": "This function supports infrastructure deployment and is not deployed inside a VPC.",
            },
            Object {
              "id": "W92",
              "reason": "This function supports infrastructure deployment and does not require setting ReservedConcurrentExecutions.",
            },
          ],
        },
      },
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters00f2465b209575c44977ca8ff1c81d86d002849272ede5a8a90ac87dfff1e494S3Bucket33B61C3E",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters00f2465b209575c44977ca8ff1c81d86d002849272ede5a8a90ac87dfff1e494S3VersionKeyDE2514FF",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters00f2465b209575c44977ca8ff1c81d86d002849272ede5a8a90ac87dfff1e494S3VersionKeyDE2514FF",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": Object {
          "Fn::Join": Array [
            "",
            Array [
              "Lambda function for auto-deleting objects in ",
              Object {
                "Ref": "SecureBucket747CD8C0",
              },
              " S3 bucket.",
            ],
          ],
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "SecureBucket747CD8C0": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "KMSMasterKeyID": Object {
                  "Fn::GetAtt": Array [
                    "SecureBucketCmkB881412D",
                    "Arn",
                  ],
                },
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "BucketName": "aws-accelerator-secure-bucket",
        "OwnershipControls": Object {
          "Rules": Array [
            Object {
              "ObjectOwnership": "BucketOwnerPreferred",
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": Object {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "SecureBucketAutoDeleteObjectsCustomResource2AF5EBAB": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "SecureBucketPolicy6374AC61",
      ],
      "Properties": Object {
        "BucketName": Object {
          "Ref": "SecureBucket747CD8C0",
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "SecureBucketCmkAliasE1748931": Object {
      "Properties": Object {
        "AliasName": "alias/accelerator/installer/s3",
        "TargetKeyId": Object {
          "Fn::GetAtt": Array [
            "SecureBucketCmkB881412D",
            "Arn",
          ],
        },
      },
      "Type": "AWS::KMS::Alias",
    },
    "SecureBucketCmkB881412D": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "Description": "AWS Accelerator Installer Bucket CMK",
        "EnableKeyRotation": true,
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kms:Create*",
                "kms:Describe*",
                "kms:Enable*",
                "kms:List*",
                "kms:Put*",
                "kms:Update*",
                "kms:Revoke*",
                "kms:Disable*",
                "kms:Get*",
                "kms:Delete*",
                "kms:ScheduleKeyDeletion",
                "kms:CancelKeyDeletion",
                "kms:GenerateDataKey",
                "kms:TagResource",
                "kms:UntagResource",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "SecureBucketPolicy6374AC61": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "SecureBucket747CD8C0",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::GetAtt": Array [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "SecureBucket747CD8C0",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "SecureBucket747CD8C0",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "s3:PutObject",
              "Condition": Object {
                "StringNotEquals": Object {
                  "s3:x-amz-server-side-encryption": "aws:kms",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "SecureBucket747CD8C0",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
              "Sid": "deny-non-encrypted-object-uploads",
            },
            Object {
              "Action": "s3:*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "SecureBucket747CD8C0",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
              "Sid": "deny-insecure-connections",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
}
`;
